### PR TITLE
ACS-7605: Initial version of handling various authentication ACS methods 

### PR DESCRIPTION
**Pull request by mpichura**
This is a first step of refactoring authentication handling in all modules and it only concerns prediction handler authentication against ACS.
There are some parts of the code which will get changed in future PRs and I'll try to explain by commenting in this PR.

### CODE REVIEW

File: common-authentication/src/main/java/org/alfresco/hxi_connector/common/adapters/auth/AuthenticationClient.java [hunk 1]
```diff
@@ -31,4 +31,9 @@ public interface AuthenticationClient
 {
 
     AuthenticationResult authenticate(String tokenUri, ClientRegistration clientRegistration);
+
+    default AuthenticationResult authenticate(String clientRegistrationId)
```
Comment (mpichura) [inline common-authentication/src/main/java/org/alfresco/hxi_connector/common/adapters/auth/AuthenticationClient.java:35]: Generally, this method is planned to be the only one used across the applications and the one above will gradually be replaced by this one.

File: prediction-applier/src/integration-test/java/org/alfresco/hxi_connector/prediction_applier/adapters/auth/PredictionApplierHxAuthClientIntegrationTest.java [hunk 2]
```diff
@@ -25,23 +25,78 @@
...
 @Testcontainers
 @SuppressWarnings("PMD.TestClassWithoutTestCases")
 class PredictionApplierHxAuthClientIntegrationTest extends HxAuthenticationClientTest
-{}
+{
+    @Container
+    private static final WireMockContainer ACS_MOCK = DockerContainers.createWireMockContainer();
+
+    @Test
+    protected void testAuthorize_alfresco()
```
Comment (mpichura) [inline prediction-applier/src/integration-test/java/org/alfresco/hxi_connector/prediction_applier/adapters/auth/PredictionApplierHxAuthClientIntegrationTest.java:72]: This method will likely be placed in the parent test class as it tests the newly added `authenticate` method.

File: prediction-applier/src/main/java/org/alfresco/hxi_connector/prediction_applier/auth/PredictionApplierHxAuthClient.java [hunk 3]
```diff
@@ -51,4 +65,44 @@ public AuthenticationResult authenticate(String tokenUri, ClientRegistration cli
     {
         return super.authenticate(tokenUri, clientRegistration);
     }
+
+    @Retryable(retryFor = EndpointServerErrorException.class,
+            maxAttemptsExpression = "#{@hxInsightProperties.hylandExperience.authentication.retry.attempts}",
+            backoff = @Backoff(delayExpression = "#{@hxInsightProperties.hylandExperience.authentication.retry.initialDelay}",
+                    multiplierExpression = "#{@hxInsightProperties.hylandExperience.authentication.retry.delayMultiplier}"))
+    @Override
+    public AuthenticationResult authenticate(String clientRegistrationId)
```
Comment (mpichura) [inline prediction-applier/src/main/java/org/alfresco/hxi_connector/prediction_applier/auth/PredictionApplierHxAuthClient.java:74]: This method will be moved to the parent class and will be simplified to trigger camel authentication rout with needed parameters.

File: prediction-applier/src/main/java/org/alfresco/hxi_connector/prediction_applier/auth/PredictionApplierHxAuthClient.java [hunk 4]
```diff
@@ -51,4 +65,44 @@ public AuthenticationResult authenticate(String tokenUri, ClientRegistration cli
...
+                .clientSecret(registration.getClientSecret())
+                .scope(registration.getScope())
+                .build();
+        return super.authenticate(tokenUri, clientRegistration);
+    }
+
+    @Override
+    protected String createEncodedBody(ClientRegistration clientRegistration)
+    {
+        if (AuthorizationGrantType.PASSWORD.getValue().equals(clientRegistration.getAuthorizationGrantType().getValue()))
```
Comment (mpichura) [inline prediction-applier/src/main/java/org/alfresco/hxi_connector/prediction_applier/auth/PredictionApplierHxAuthClient.java:94]: This is special handling of `password` grant type which is not supported by Spring Boot OAuth2 auto configuration.

**Review by Pawel-608 — APPROVED** [review]
LGTM

File: common-authentication/src/main/java/org/alfresco/hxi_connector/common/adapters/auth/TokenRequest.java [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+
+        return body.toString();
+    }
+
+    public static TokenRequestBuilder builder()
+    {
+        return new TokenRequestBuilder();
+    }
+
+    public static class TokenRequestBuilder
```
[the remark this answers is not in the document]
Comment (mpichura) [reply common-authentication/src/main/java/org/alfresco/hxi_connector/common/adapters/auth/TokenRequest.java:78]: In some cases, I need to ignore/skip blank values. If you happen to know any lombok magic that could help here - I am happy to use that.

File: prediction-applier/src/main/java/org/alfresco/hxi_connector/prediction_applier/auth/PredictionApplierHxAuthClient.java [hunk 6]
```diff
@@ -25,21 +25,35 @@
...
+import org.alfresco.hxi_connector.common.adapters.auth.TokenRequest;
 import org.alfresco.hxi_connector.common.exception.EndpointServerErrorException;
 import org.alfresco.hxi_connector.prediction_applier.config.HxInsightProperties;
+import org.alfresco.hxi_connector.prediction_applier.config.NodesApiProperties;
 
+@Component
 public class PredictionApplierHxAuthClient extends HxAuthenticationClient
 {
-    public PredictionApplierHxAuthClient(CamelContext camelContext, HxInsightProperties hxInsightProperties)
+    private final OAuth2ClientProperties oAuth2ClientProperties;
```
[the remark this answers is not in the document]
Comment (mpichura) [reply prediction-applier/src/main/java/org/alfresco/hxi_connector/prediction_applier/auth/PredictionApplierHxAuthClient.java:48]: That could do. Once Spring Security is replaced with custom auth handling, we can think about custom auth properties handling as well.

**Pull request merged**